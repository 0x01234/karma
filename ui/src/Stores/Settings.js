import { localStored } from "mobx-stored";

import { action } from "mobx";

const defaultSavedFilters = {
  filters: [],
  present: false
};

class Settings {
  savedFilters = localStored("savedFilters", defaultSavedFilters, {
    delay: 100
  });

  saveFilters = action(newFilters => {
    this.savedFilters.filters = newFilters;
    this.savedFilters.present = true;
  });

  clearSavedFilters = action(() => {
    this.savedFilters.filters = [];
    this.savedFilters.present = false;
  });
}

export { Settings };
