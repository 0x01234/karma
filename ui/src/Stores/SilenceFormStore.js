import { observable, action, computed } from "mobx";

import uniqueId from "lodash.uniqueid";

import moment from "moment";

const NewEmptyMatcher = id => {
  return {
    id: uniqueId(),
    name: "",
    values: [],
    suggestions: {
      names: [],
      values: []
    },
    isRegex: false
  };
};

const ValueToObject = value => ({ label: value, value: value });

class SilenceFormStore {
  // this is used to store modal visibility toggle
  toggle = observable(
    {
      visible: false,
      toggle() {
        this.visible = !this.visible;
      },
      hide() {
        this.visible = false;
      },
      show() {
        this.visible = true;
      }
    },
    { toggle: action.bound, hide: action.bound, show: action.bound }
  );

  // form data is stored here, it's global (rather than attached to the form)
  // so it can be manipulated from other parts of the code
  // example: when user clicks a silence button on alert we should populate
  // this form from that alert so user can easily silence that alert
  data = observable(
    {
      inProgress: false,
      alertmanagers: [],
      matchers: [],
      startsAt: moment(),
      endsAt: moment().add(1, "hour"),
      comment: "",
      author: "",

      resetProgress() {
        this.inProgress = false;
      },

      // append a new empty matcher to the list
      addEmptyMatcher() {
        let m = NewEmptyMatcher();
        this.matchers.push(m);
      },

      deleteMatcher(id) {
        // only delete matchers if we have more than 1
        if (this.matchers.length > 1) {
          this.matchers = this.matchers.filter(m => m.id !== id);
        }
      },

      fillMatchersFromGroup(group) {
        let matchers = [];

        // add matchers for all shared labels in this group
        for (const [key, value] of Object.entries(
          Object.assign({}, group.labels, group.shared.labels)
        )) {
          matchers.push({
            id: uniqueId(),
            name: key,
            values: [ValueToObject(value)],
            suggestions: {
              names: [],
              values: []
            },
            isRegex: false
          });
        }

        // add matchers for all unique labels in this group
        let labels = {};
        for (const alert of group.alerts) {
          for (const [key, value] of Object.entries(alert.labels)) {
            if (!labels[key]) {
              labels[key] = new Set();
            }
            labels[key].add(value);
          }
        }
        for (const [key, values] of Object.entries(labels)) {
          matchers.push({
            id: uniqueId(),
            name: key,
            values: [...values].sort().map(value => ValueToObject(value)),
            suggestions: {
              names: [],
              values: []
            },
            isRegex: values.size > 1
          });
        }

        this.matchers = matchers;
      },

      verifyStarEnd() {
        if (this.endsAt.isSameOrBefore(this.startsAt)) {
          this.endsAt = moment(this.startsAt).add(1, "minutes");
        }
      },
      incStart(minutes) {
        this.startsAt = moment(this.startsAt).add(minutes, "minutes");
        this.verifyStarEnd();
      },
      decStart(minutes) {
        this.startsAt = moment(this.startsAt).subtract(minutes, "minutes");
        this.verifyStarEnd();
      },

      incEnd(minutes) {
        this.endsAt = moment(this.endsAt).add(minutes, "minutes");
        this.verifyStarEnd();
      },
      decEnd(minutes) {
        this.endsAt = moment(this.endsAt).subtract(minutes, "minutes");
        this.verifyStarEnd();
      },

      incDuration(minutes) {
        this.endsAt = moment(this.endsAt).add(minutes, "minutes");
      },
      decDuration(minutes) {
        const newEndsAt = moment(this.endsAt).subtract(minutes, "minutes");
        if (newEndsAt.isAfter(this.startsAt)) {
          this.endsAt = newEndsAt;
        }
      },

      get toAlertmanagerPayload() {
        const payload = {
          matchers: this.matchers.map(m => ({
            name: m.name,
            value:
              m.values.length > 1
                ? `(${m.values.map(v => v.value).join("|")})`
                : m.values.length === 1
                  ? m.values[0].value
                  : "",
            isRegex: m.isRegex
          })),
          startsAt: this.startsAt
            .second(0)
            .millisecond(0)
            .toISOString(),
          endsAt: this.endsAt
            .second(59)
            .millisecond(0)
            .toISOString(),
          createdBy: this.author,
          comment: this.comment
        };
        return payload;
      },
      get toDuration() {
        const data = {
          days: this.endsAt.diff(this.startsAt, "days"),
          hours: this.endsAt.diff(this.startsAt, "hours") % 24,
          minutes: this.endsAt.diff(this.startsAt, "minutes") % 60
        };
        return data;
      }
    },
    {
      resetProgress: action.bound,
      addEmptyMatcher: action.bound,
      deleteMatcher: action.bound,
      fillMatchersFromGroup: action.bound,
      verifyStarEnd: action.bound,
      incStart: action.bound,
      decStart: action.bound,
      incEnd: action.bound,
      decEnd: action.bound,
      incDuration: action.bound,
      decDuration: action.bound,
      toAlertmanagerPayload: computed,
      toDuration: computed
    },
    { name: "Silence form store" }
  );
}

export { SilenceFormStore };
