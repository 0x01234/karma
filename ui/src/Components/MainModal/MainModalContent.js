import React, { Component } from "react";
import ReactDOM from "react-dom";
import PropTypes from "prop-types";

import { observer } from "mobx-react";
import { observable, action } from "mobx";

import { disableBodyScroll, enableBodyScroll } from "body-scroll-lock";

import { AlertStore } from "Stores/AlertStore";
import { Settings } from "Stores/Settings";
import { Configuration } from "./Configuration";
import { MountModalBackdrop } from "Components/Animations/MountModal";
import { Help } from "./Help";

const Tab = ({ title, active, onClick }) => (
  <span
    className={`nav-item nav-link cursor-pointer ${
      active ? "active" : "text-primary"
    }`}
    onClick={onClick}
  >
    {title}
  </span>
);
Tab.propTypes = {
  title: PropTypes.string.isRequired,
  active: PropTypes.bool,
  onClick: PropTypes.func.isRequired
};

const TabNames = Object.freeze({
  Configuration: "configuration",
  Help: "help"
});

const MainModalContent = observer(
  class MainModalContent extends Component {
    static propTypes = {
      alertStore: PropTypes.instanceOf(AlertStore).isRequired,
      settingsStore: PropTypes.instanceOf(Settings).isRequired,
      onHide: PropTypes.func.isRequired,
      isVisible: PropTypes.bool.isRequired
    };

    tab = observable(
      {
        current: TabNames.Configuration,
        setTab(newTab) {
          this.current = newTab;
        }
      },
      { setTab: action.bound }
    );

    componentDidMount() {
      disableBodyScroll(document.querySelector(".modal"));
    }

    componentWillUnmount() {
      enableBodyScroll(document.querySelector(".modal"));
    }

    render() {
      const { alertStore, settingsStore, onHide, isVisible } = this.props;

      return ReactDOM.createPortal(
        <React.Fragment>
          <div className="modal d-block" role="dialog">
            <div className="modal-dialog modal-lg" role="document">
              <div className="modal-content">
                <div className="modal-header py-2">
                  <nav className="nav nav-pills nav-justified w-100">
                    <Tab
                      title="Configuration"
                      active={this.tab.current === TabNames.Configuration}
                      onClick={() => this.tab.setTab(TabNames.Configuration)}
                    />
                    <Tab
                      title="Help"
                      active={this.tab.current === TabNames.Help}
                      onClick={() => this.tab.setTab(TabNames.Help)}
                    />
                    <button type="button" className="close" onClick={onHide}>
                      <span>&times;</span>
                    </button>
                  </nav>
                </div>
                <div className="modal-body">
                  {this.tab.current === TabNames.Help ? <Help /> : null}
                  {this.tab.current === TabNames.Configuration ? (
                    <Configuration settingsStore={settingsStore} />
                  ) : null}
                </div>
                <div className="modal-footer">
                  <span className="text-muted">
                    Version: {alertStore.info.version}
                  </span>
                </div>
              </div>
            </div>
          </div>
          <MountModalBackdrop in={isVisible} unmountOnExit>
            <div className="modal-backdrop d-block" />
          </MountModalBackdrop>
        </React.Fragment>,
        document.body
      );
    }
  }
);

export { MainModalContent };
