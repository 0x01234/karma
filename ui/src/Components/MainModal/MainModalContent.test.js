import React from "react";

import { mount } from "enzyme";

import toDiffableHtml from "diffable-html";

import { AlertStore } from "Stores/AlertStore";
import { Settings } from "Stores/Settings";
import { MainModalContent } from "./MainModalContent";

let alertStore;
let settingsStore;
const onHide = jest.fn();

beforeEach(() => {
  alertStore = new AlertStore([]);
  settingsStore = new Settings();
  onHide.mockClear();
});

const FakeModal = () => {
  return mount(
    <MainModalContent
      alertStore={alertStore}
      settingsStore={settingsStore}
      onHide={onHide}
    />
  );
};

const ValidateSetTab = (title, callArg) => {
  const component = FakeModal();
  const instance = component.instance();
  const setTabSpy = jest.spyOn(instance.tab, "setTab");

  const tab = component.find({ title: title });
  tab.simulate("click");
  expect(setTabSpy).toHaveBeenCalledWith(callArg);
  expect(component.find(".nav-link.active").text()).toBe(title);
};

describe("<MainModalContent />", () => {
  it("matches snapshot", () => {
    const tree = FakeModal();
    expect(toDiffableHtml(tree.html())).toMatchSnapshot();
  });

  it("shows 'Configuration' tab by default", () => {
    const tree = FakeModal();
    const activeTab = tree.find(".nav-link.active");
    expect(activeTab.text()).toBe("Configuration");
  });

  // modal makes it tricky to verify re-rendered content, so only check if we
  // update the store for now
  it("calls setTab('configuration') after clicking on the 'Configuration' tab", () => {
    ValidateSetTab("Configuration", "configuration");
  });

  it("calls setTab('help') after clicking on the 'Help' tab", () => {
    ValidateSetTab("Help", "help");
  });
});
