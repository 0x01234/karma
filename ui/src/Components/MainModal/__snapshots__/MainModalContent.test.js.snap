// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MainModalContent /> matches snapshot 1`] = `
"
<div class=\\"modal d-block\\"
     role=\\"dialog\\"
>
  <div class=\\"modal-dialog modal-lg\\"
       role=\\"document\\"
  >
    <div class=\\"modal-content\\">
      <div class=\\"modal-header py-2\\">
        <nav class=\\"nav nav-pills nav-justified w-100\\">
          <span class=\\"nav-item nav-link cursor-pointer active\\">
            Configuration
          </span>
          <span class=\\"nav-item nav-link cursor-pointer text-primary\\">
            Help
          </span>
          <button type=\\"button\\"
                  class=\\"close\\"
          >
            <span>
              Ã—
            </span>
          </button>
        </nav>
      </div>
      <div class=\\"modal-body\\">
        <form class=\\"px-3\\">
          <div class=\\"form-group text-center\\">
            <label class=\\"mb-4\\">
              Refresh interval
            </label>
            <div aria-disabled=\\"false\\"
                 class=\\"input-range\\"
            >
              <span class=\\"input-range__label input-range__label--min\\">
                <span class=\\"input-range__label-container\\">
                  10s
                </span>
              </span>
              <div class=\\"input-range__track input-range__track--background\\">
                <div style=\\"left: 0%; width: 18.181818181818183%;\\"
                     class=\\"input-range__track input-range__track--active\\"
                >
                </div>
                <span class=\\"input-range__slider-container\\"
                      style=\\"position: absolute; left: 18.181818181818183%;\\"
                >
                  <span class=\\"input-range__label input-range__label--value\\">
                    <span class=\\"input-range__label-container\\">
                      30s
                    </span>
                  </span>
                  <div aria-valuemax=\\"120\\"
                       aria-valuemin=\\"10\\"
                       aria-valuenow=\\"30\\"
                       class=\\"input-range__slider\\"
                       draggable=\\"false\\"
                       role=\\"slider\\"
                       tabindex=\\"0\\"
                  >
                  </div>
                </span>
              </div>
              <span class=\\"input-range__label input-range__label--max\\">
                <span class=\\"input-range__label-container\\">
                  120s
                </span>
              </span>
            </div>
          </div>
          <div class=\\"mt-5\\">
          </div>
          <div class=\\"form-group text-center\\">
            <label class=\\"mb-4\\">
              Default number of alerts to show per group
            </label>
            <div aria-disabled=\\"false\\"
                 class=\\"input-range\\"
            >
              <span class=\\"input-range__label input-range__label--min\\">
                <span class=\\"input-range__label-container\\">
                  1
                </span>
              </span>
              <div class=\\"input-range__track input-range__track--background\\">
                <div style=\\"left: 0%; width: 44.44444444444444%;\\"
                     class=\\"input-range__track input-range__track--active\\"
                >
                </div>
                <span class=\\"input-range__slider-container\\"
                      style=\\"position: absolute; left: 44.44444444444444%;\\"
                >
                  <span class=\\"input-range__label input-range__label--value\\">
                    <span class=\\"input-range__label-container\\">
                      5
                    </span>
                  </span>
                  <div aria-valuemax=\\"10\\"
                       aria-valuemin=\\"1\\"
                       aria-valuenow=\\"5\\"
                       class=\\"input-range__slider\\"
                       draggable=\\"false\\"
                       role=\\"slider\\"
                       tabindex=\\"0\\"
                  >
                  </div>
                </span>
              </div>
              <span class=\\"input-range__label input-range__label--max\\">
                <span class=\\"input-range__label-container\\">
                  10
                </span>
              </span>
            </div>
          </div>
        </form>
      </div>
      <div class=\\"modal-footer\\">
        <span class=\\"text-muted\\">
          Version: unknown
        </span>
      </div>
    </div>
  </div>
</div>
"
`;
